endpoints/events/event.go

```diff
@@ -0,0 +1,324 @@
+	}
+
+	return ee.Handle
+}
+
+func (e *eventEndpoint) Handle(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	// parse event request from http req
+	eventRequest, errs := ParseEventRequest(r)
+
+	// handle possible parsing errors
+	if len(errs) > 0 {
+		w.WriteHeader(http.StatusBadRequest)
+
+		for _, err := range errs {
+			w.Write([]byte(fmt.Sprintf("invalid request: %s\n", err.Error())))
+		}
+
+		return
+	}
+
+	// validate account id
+	accountId, err := validateRequiredParameter(r, AccountIdParameter)
+
+	if err != nil {
+		w.WriteHeader(http.StatusUnauthorized)
+		w.Write([]byte(fmt.Sprintf("Account '%s' is required query parameter and can't be empty", AccountIdParameter)))
+		return
+	}
+	eventRequest.AccountID = accountId
+
+	if eventRequest.Analytics != analytics.Enabled {
+		return
+	}
+
+	ctx := context.Background()
+
+	// get account details
+	account, errs := accountService.GetAccount(ctx, e.Cfg, e.Accounts, eventRequest.AccountID)
+	if len(errs) > 0 {
+		status, messages := HandleAccountServiceErrors(errs)
+		w.WriteHeader(status)
+
+		for _, message := range messages {
+			w.Write([]byte(fmt.Sprintf("Invalid request: %s\n", message)))
+		}
+		return
+	}
+
+	// account does not support events
+	if !account.EventsEnabled {
+		w.WriteHeader(http.StatusUnauthorized)
+		w.Write([]byte(fmt.Sprintf("Account '%s' doesn't support events", eventRequest.AccountID)))
+		return
+	}
+
+	// handle notification event
+	e.Analytics.LogNotificationEventObject(&analytics.NotificationEvent{
+		Request: eventRequest,
+		Account: account,
+	})
+
+	// OK
+	w.WriteHeader(http.StatusOK)
+	// Add tracking pixel if format == image
+	if eventRequest.Format == analytics.Image {
+		w.Header().Add("Content-Type", e.TrackingPixel.ContentType)
+		w.Write(e.TrackingPixel.Content)
+	}
+}
+
+// Converts an EventRequest to an URL
+func EventRequestToUrl(externalUrl string, request *analytics.EventRequest) string {
+	s := fmt.Sprintf(TemplateUrl, externalUrl, request.Type, request.Bidid, request.AccountID)
+
+	return s + optionalParameters(request)
+}
+
+// Parses an EventRequest from an Http request
+func ParseEventRequest(r *http.Request) (*analytics.EventRequest, []error) {
+	event := &analytics.EventRequest{}
+	var errs []error
+	// validate type
+	if err := validateType(event, r); err != nil {
+		errs = append(errs, err)
+	}
+
+	// validate bidid
+	if bidid, err := validateRequiredParameter(r, BidIdParameter); err != nil {
+		errs = append(errs, err)
+	} else {
+		event.Bidid = bidid
+	}
+
+	// validate timestamp (optional)
+	if err := validateTimestamp(event, r); err != nil {
+		errs = append(errs, err)
+	}
+
+	// validate format (optional)
+	if err := validateFormat(event, r); err != nil {
+		errs = append(errs, err)
+	}
+
+	// validate analytics (optional)
+	if err := validateAnalytics(event, r); err != nil {
+		errs = append(errs, err)
+	}
+
+	// Bidder
+	bidder := r.FormValue(BidderParameter)
+
+	if bidder != "" {
```

Comment:
Could this be simplified to `event.Bidder = r.URL.Query().Get(BidderParameter)`? Even if this returns an empty string, I don't see the harm of making the assignment since it defaults to an empty string anyway. Not sure the complexity of the extra `if` statement is worth it.